### PR TITLE
Add Write support to OCP/ICalendar

### LINKED ISSUE #20154: PHP Calendar API: Write support
`OCP\Calendar\IManager` is currently only providing a read-only API.
This is drastically limiting the number of use-cases, forcing app-developers to use the CalDAV interface via javascript, if they desire write-support for the user's calendar.

Nextcloud should provide a PHP api with write-support for the user's calendar.

Due to it's simple design, our current API is limited to read-only. Adding write-support to the existing API is not possible. This is mostly caused by the structure of events.
They are returned as a very simple key-value array. Some information from the ICS is already lost when the conversion takes place. Converting modified key-value arrays back to ICS could possibly lead to data-loss.

---

Proposal:
1. Deprecate `\OCP\Calendar\IManager` and `\OCP\Calendar\ICalendar` with Nextcloud 20 and remove with Nextcloud 23.

2.
- `Implement` `\OCP\Calendar\IManagerV2`:
```php
interface \OCP\Calendar\IManagerV2{
        public function isEnabled():bool;
        public function getCalendars():ICalendarV2[];
        public function getCalendar(string $key):?ICalendarV2;
        public function search(string $pattern, array $searchProperties=[], array $options=[], int $limit=null, int $offset=null)?ICalendarObjectV2[];
        public function registerCalendar(ICalendarV2 $calendar):void;
        public function unregisterCalendar(ICalendarV2 $calendar):void;
        public function register(\Closure $callable):void;
        public function clear():void;
}
```

- `Implement` `\OCP\Calendar\ICalendarV2`:
```php
interface \OCP\Calendar\ICalendarV2{
	public function getKey():string;
	public function getDisplayName():?string;
	public function getDisplayColor():?string;
        public function isWriteable():bool;
        public function getByUri(string $uri):?ICalendarObjectV2;
        public function search(string $pattern, array $searchProperties=[], array $options=[], int $limit=null, int $offset=null)?ICalendarObjectV2[];
        public function create(\Sabre\VObject\VCalendar $vObject):ICalendarObjectV2;
}
```

- `Implement` `\OCP\Calendar\ICalendarObjectV2`:
```php
interface \OCP\Calendar\ICalendarObjectV2{
        public function getCalendarKey():string;
        public function getUri():string;
        public function getVObject():\Sabre\VObject\VCalendar;
        public function update():void;
        public function delete():void;
}
```

### PR DESCRIPTION
**Pull request by tcitworld**
Fixes #20154

### CODE REVIEW

File: apps/dav/lib/CalDAV/CalendarObjectImplV2.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	/** @var array */
+	private $calendarObjectData;
+
+	/**
+	 * CalendarImpl constructor.
+	 *
+	 * @param array $calendarObjectData
+	 * @param CalDavBackend $backend
+	 */
+	public function __construct(array $calendarObjectData, CalDavBackend $backend) {
```
Comment (tcitworld) [inline apps/dav/lib/CalDAV/CalendarObjectImplV2.php]: Should be directly a `VCALENDAR` object instead.

Comment (tcitworld) [inline apps/dav/lib/CalDAV/CalendarObjectImplV2.php]: Or `__construct($calendarId, $uri, VCalendar $data, CalDavBackend $backend)` ?

File: apps/dav/lib/CalDAV/CalendarObjectImplV2.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	public function getUri(): string {
+		return $this->calendarObjectData['uri'];
+	}
+
+	public function getVObject(): VCalendar {
+		return Reader::read($this->calendarObjectData['calendardata']);
+	}
+
+	public function update(VCalendar $data): void {
+		$this->backend->updateCalendarObject($this->getCalendarKey(), $this->getUri(), $data->serialize());
```
Comment (tcitworld) [inline apps/dav/lib/CalDAV/CalendarObjectImplV2.php]: Needs VCalendar validation.

File: apps/dav/lib/CalDAV/CalendarImplV2.php [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+	/**
+	 * @param VCalendar $vObject
+	 * @return ICalendarObjectV2
+	 * @throws BadRequest
+	 * @since 19.0.0
+	 */
+	public function create(VCalendar $vObject): ICalendarObjectV2 {
+		$uuid = UUIDUtil::getUUID();
+		$this->backend->createCalendarObject($this->getKey(), $uuid, $vObject->serialize());
```
Comment (tcitworld) [inline apps/dav/lib/CalDAV/CalendarImplV2.php]: Needs VCalendar validation.

File: apps/dav/lib/CalDAV/CalendarObjectImplV2.php (hunk 1 above)
Comment (tcitworld) [reply apps/dav/lib/CalDAV/CalendarObjectImplV2.php]: What's your opinion @georgehrke ?

**Comment by georgehrke** [comment]
Moved to Nc 20 milestone, since we are past feature freeze

File: apps/dav/lib/CalDAV/CalendarObjectImplV2.php (hunk 1 above)
Comment (georgehrke) [reply apps/dav/lib/CalDAV/CalendarObjectImplV2.php:45]: I feel like `__construct($calendarId, $uri, VCalendar $data, CalDavBackend $backend)` is cleaner. Then we can also properly enforce types $calendarId and $uri.

Comment (tcitworld) [reply apps/dav/lib/CalDAV/CalendarObjectImplV2.php]: Done. Feels a bit odd to have $calendarId as a `int` type but `calendarKey` will be given as a string though.

**Comment by MorrisJobke** [comment]
> Moved to Nc 20 milestone, since we are past feature freeze

Unfortunately we are close to miss this again. I feel also a bit uncomfortable with adding a that big change that late.

**Comment by rotdrop** [comment]
FYI, I have rebased the branch of @tcitworld and submitted the merge request #23234 (mostly in order to get the merge checks run). Would be great to have this integrated.

Other things I still miss are
- refresh the list of calendars after addings or removing or sharing calenders by whatever means
- create/delete calendars
- share/unshare calendars

But even without those 3 points the pull request would reduce the need to access internals of the dav-app.

**Pull request closed without merging**